# Principal Component Analysis

In [20]:
from sklearn.decomposition import PCA
import pandas as pd
iris = pd.read_csv('Iris.csv')
X = iris.drop(['Id','Class Label'], axis=1)
#Y = iris.drop('Id',axis=1)
pca = PCA(n_components=4)
pca.fit(X)
#pca.fit(Y)

PCA(n_components=4)

In [21]:
print(pd.DataFrame(pca.components_,columns=X.columns))

   SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
0       0.361590     -0.082269       0.856572      0.358844
1       0.656540      0.729712      -0.175767     -0.074706
2      -0.580997      0.596418       0.072524      0.549061
3       0.317255     -0.324094      -0.479719      0.751121


# Based on these coefficients, we can calculate the PCA components for our input DataFrame X:

In [22]:
pca_df=(pd.DataFrame(pca.components_,columns=X.columns))

# Let us calculate PC1 using coefficients that are generated
X['PC1'] = X['SepalLengthCm']* pca_df['SepalLengthCm'][0] + X['SepalWidthCm']*pca_df['SepalWidthCm'][0]+ X['PetalLengthCm']*pca_df['PetalLengthCm'][0]+X['PetalWidthCm']* pca_df['PetalWidthCm'][0]

# Let us calculate PC2
X['PC2'] = X['SepalLengthCm']* pca_df['SepalLengthCm'][1] + X['SepalWidthCm']*pca_df['SepalWidthCm'][1]+ X['PetalLengthCm']*pca_df['PetalLengthCm'][1]+X['PetalWidthCm']* pca_df['PetalWidthCm'][1]

#Let us calculate PC3
X['PC3'] = X['SepalLengthCm']* pca_df['SepalLengthCm'][2] + X['SepalWidthCm']*pca_df['SepalWidthCm'][2]+ X['PetalLengthCm']*pca_df['PetalLengthCm'][2]+X['PetalWidthCm']* pca_df['PetalWidthCm'][2]

# Let us calculate PC4
X['PC4'] = X['SepalLengthCm']* pca_df['SepalLengthCm'][3] + X['SepalWidthCm']*pca_df['SepalWidthCm'][3]+ X['PetalLengthCm']*pca_df['PetalLengthCm'][3]+X['PetalWidthCm']* pca_df['PetalWidthCm'][3]

In [23]:
print(X)

     SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm       PC1  \
0              5.1           3.5            1.4           0.2  2.827136   
1              4.9           3.0            1.4           0.2  2.795952   
2              4.7           3.2            1.3           0.2  2.621524   
3              4.6           3.1            1.5           0.2  2.764906   
4              5.0           3.6            1.4           0.2  2.782750   
..             ...           ...            ...           ...       ...   
145            6.7           3.0            5.2           2.3  7.455360   
146            6.3           2.5            5.0           1.9  7.037007   
147            6.5           3.0            5.2           2.0  7.275389   
148            6.2           3.4            5.4           2.3  7.412972   
149            5.9           3.0            5.1           1.8  6.901009   

          PC2       PC3       PC4  
0    5.641331 -0.664277 -0.037715  
1    5.145167 -0.846287  0.

In [24]:
print(pca.explained_variance_ratio_)

[0.92461621 0.05301557 0.01718514 0.00518309]


# The variance ratio indicates the following:
If we choose to replace the original four features with PC1, then we will be able to capture about 92.4% of the variance of the original variables. We will introduce some approximations by not capturing 100% of the variance of the original four
features.

If we choose to replace the original four features with PC1 and PC2, then we will capture an additional 5.3 % of the variance of the original variables.

If we choose to replace the original four features with PC1, PC2, and PC3, then we will now capture a further 0.017 % of the variance of the original variables.

If we choose to replace the original four features with four principal components, then we will capture 100% of the variance of the original variables (92.4 + 0.053 + 0.017 + 0.005), but replacing four original features with four principal components is meaningless as we did not reduce the dimensions at all and achieved nothing.